backend/tournesol/views/inconsistencies.py

```diff
@@ -8,12 +8,188 @@
+    all_connections = {}
+
+    def get(self, request, *args, **kwargs):
+        poll = self.poll_from_url
+        user = request.user
+
+        stats = {}
+        self.cycles = []
+
+        filter_serializer = ScoreInconsistenciesFilterSerializer(data=request.query_params)
+        filter_serializer.is_valid(raise_exception=True)
+        filters = filter_serializer.validated_data
+        comparisons = ComparisonCriteriaScore.objects.filter(
+            comparison__user=user,
+            comparison__poll=poll,
+        )
+
+        if filters["date_gte"]:
+            comparisons = comparisons.filter(
+                comparison__datetime_lastedit__gte=filters["date_gte"]
+            )
+
+        comparisons = list(
+            comparisons.values(
+                "comparison__entity_1__uid",
+                "comparison__entity_2__uid",
+                "criteria",
+                "score",
+            )
+        )
+
+        for self.criterion in poll.criterias_list:
+
+            self._fill_graph_parameters(comparisons)
+
+            cycles_count, comparison_trios_count = self._count_cycles_and_comparison_trios()
+
+            stats[self.criterion] = {
+                "cycles_count": cycles_count,
+                "comparison_trios_count": comparison_trios_count,
+            }
+
+        response = {
+            "count": len(self.cycles),
+            "results": self.cycles,
+            "stats": stats,
+        }
+
+        return Response(Length3CyclesSerializer(response).data)
+
+    def _fill_graph_parameters(self, comparisons: list):
+        """
+        Fill the set containing all the entities.
+
+        The incoming and outgoing connections sets are used to search the cycles.
+        all_connections is only used to count all the comparison trios.
+
+        outgoing_connections: Dictionary of sets for each entity.
+                              Each element in the sets is a preferred entity.
+        incoming_connections: Idem but each element in the sets is a worse rated entity.
+
+        all_connections: all the edges (also including comparisons with a score of 0 as edges)
+        """
+        self.all_entities = set()
+        self.outgoing_connections = {}
+        self.incoming_connections = {}
+
+        # all the edges (also including comparisons with a score of 0 as edges)
+        self.all_connections = {}
+
+        criterion_comparisons = list(
+            filter(lambda d: d["criteria"] == self.criterion, comparisons)
+        )
+
+        for comparison in criterion_comparisons:
+            entity_1 = comparison["comparison__entity_1__uid"]
+            entity_2 = comparison["comparison__entity_2__uid"]
+
+            if entity_1 not in self.all_entities:
+                self.all_entities.add(entity_1)
+                self.incoming_connections[entity_1] = set()
+                self.outgoing_connections[entity_1] = set()
+                self.all_connections[entity_1] = set()
```

Comment:
Using `defaultdict` would simplify the code as you would not need to initialize all dictionaries.

Reply:
agreed